aws/resource_aws_cognito_user_pool.go

```diff
@@ -0,0 +1,902 @@
+package aws
+
+import (
+	"errors"
+	"log"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	"github.com/aws/aws-sdk-go/service/cognitoidentityprovider"
+	"github.com/hashicorp/errwrap"
+	"github.com/hashicorp/terraform/helper/resource"
+	"github.com/hashicorp/terraform/helper/schema"
+	"github.com/hashicorp/terraform/helper/validation"
+)
+
+func resourceAwsCognitoUserPool() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceAwsCognitoUserPoolCreate,
+		Read:   resourceAwsCognitoUserPoolRead,
+		Update: resourceAwsCognitoUserPoolUpdate,
+		Delete: resourceAwsCognitoUserPoolDelete,
+
+		Importer: &schema.ResourceImporter{
+			State: schema.ImportStatePassthrough,
+		},
+
+		// https://docs.aws.amazon.com/cognito-user-identity-pools/latest/APIReference/API_CreateUserPool.html
+		Schema: map[string]*schema.Schema{
+			"admin_create_user_config": {
+				Type:     schema.TypeList,
+				Optional: true,
+				Computed: true,
+				MinItems: 0,
+				MaxItems: 1,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"allow_admin_create_user_only": {
+							Type:     schema.TypeBool,
+							Optional: true,
+						},
+						"invite_message_template": {
+							Type:     schema.TypeList,
+							Optional: true,
+							MinItems: 0,
+							MaxItems: 1,
+							Elem: &schema.Resource{
+								Schema: map[string]*schema.Schema{
+									"email_message": {
+										Type:         schema.TypeString,
+										Optional:     true,
+										ValidateFunc: validateCognitoUserPoolInviteTemplateEmailMessage,
+									},
+									"email_subject": {
+										Type:         schema.TypeString,
+										Optional:     true,
+										ValidateFunc: validateCognitoUserPoolTemplateEmailSubject,
+									},
+									"sms_message": {
+										Type:         schema.TypeString,
+										Optional:     true,
+										ValidateFunc: validateCognitoUserPoolInviteTemplateSmsMessage,
+									},
+								},
+							},
+						},
+						"unused_account_validity_days": {
+							Type:         schema.TypeInt,
+							Optional:     true,
+							Default:      7,
+							ValidateFunc: validateIntegerInRange(0, 90),
+						},
+					},
+				},
+			},
+
+			"alias_attributes": {
+				Type:     schema.TypeList,
```

Comment:
I assume the ordering is important here?

Reply:
Since you write it down, yeh it seems so!